File: server/services/deepseek-client.mjs
const DEFAULT_BASE_URL = "https://api.deepseek.com";
const DEFAULT_MODEL = "deepseek-v4-flash";

export class DeepSeekError extends Error {
  constructor(message, { code = "DEEPSEEK_REQUEST_FAILED", status = 502, cause } = {}) {
    super(message, { cause });
    this.name = "DeepSeekError";
    this.code = code;
    this.status = status;
  }
}

export function deepSeekConfig() {
  const testMode = process.env.NODE_ENV === "test";
  return {
    configured: Boolean(process.env.DEEPSEEK_API_KEY) && (!testMode || process.env.DEEPSEEK_ALLOW_IN_TESTS === "true"),
    baseUrl: String(process.env.DEEPSEEK_BASE_URL ?? DEFAULT_BASE_URL).replace(/\/+$/, ""),
    model: process.env.DEEPSEEK_MODEL ?? DEFAULT_MODEL
  };
}

function parseJsonContent(content) {
  const clean = String(content ?? "").trim().replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  if (!clean) throw new DeepSeekError("DeepSeek returned an empty JSON response", { code: "DEEPSEEK_EMPTY_RESPONSE" });
  try {
    return JSON.parse(clean);
  } catch (cause) {
    throw new DeepSeekError("DeepSeek returned malformed JSON", { code: "DEEPSEEK_INVALID_JSON", cause });
  }
}

export async function deepSeekChat({
  messages,
  json = false,
  maxTokens = 900,
  temperature = 1,
  userId = "aihub_mvp_user",
  timeoutMs = 40_000
}) {
  const config = deepSeekConfig();
  if (!config.configured) {
    throw new DeepSeekError("DEEPSEEK_API_KEY is not configured", {
      code: "DEEPSEEK_NOT_CONFIGURED",
      status: 503
    });
  }

  let response;
  try {
    response = await fetch(`${config.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${process.env.DEEPSEEK_API_KEY}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        model: config.model,
        messages,
        max_tokens: Math.max(64, Math.min(4_000, Number(maxTokens) || 900)),
        temperature,
        stream: false,
        user_id: String(userId).replace(/[^a-zA-Z0-9_-]/g, "_").slice(0, 128),
        ...(json ? { response_format: { type: "json_object" } } : {})
      }),
      signal: AbortSignal.timeout(timeoutMs)
    });
  } catch (cause) {
    throw new DeepSeekError("Unable to connect to DeepSeek", {
      code: cause?.name === "TimeoutError" ? "DEEPSEEK_TIMEOUT" : "DEEPSEEK_NETWORK_ERROR",
      status: 502,
      cause
    });
  }

  const raw = await response.text();
  let payload;
  try {
    payload = raw ? JSON.parse(raw) : {};
  } catch {
    payload = {};
  }
  if (!response.ok) {
    throw new DeepSeekError(payload?.error?.message || `DeepSeek request failed with ${response.status}`, {
      code: "DEEPSEEK_UPSTREAM_ERROR",
      status: response.status === 429 ? 429 : 502
    });
  }

  const content = payload?.choices?.[0]?.message?.content;
  if (!String(content ?? "").trim()) {
    throw new DeepSeekError("DeepSeek returned an empty response", { code: "DEEPSEEK_EMPTY_RESPONSE" });
  }
  return {
    content: json ? parseJsonContent(content) : String(content).trim(),
    model: payload.model ?? config.model,
    usage: payload.usage ?? null,
    finishReason: payload?.choices?.[0]?.finish_reason ?? null
  };
}

File: app/llm/deepseek_client.py
"""OpenAI-compatible chat completion client for DeepSeek."""

from __future__ import annotations

import os

import aiohttp

from app.llm.base import LLMBackend, LLMError


class DeepSeekBackend(LLMBackend):
    def __init__(
        self,
        *,
        base_url: str = "https://api.deepseek.com",
        model: str = "deepseek-v4-flash",
        api_key: str = "",
        api_key_env: str = "DEEPSEEK_API_KEY",
        temperature: float = 1.0,
        max_tokens: int = 900,
        timeout_seconds: float = 40.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.api_key = api_key.strip()
        self.api_key_env = api_key_env
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.timeout_seconds = timeout_seconds
        self._session: aiohttp.ClientSession | None = None

    def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                timeout=aiohttp.ClientTimeout(total=self.timeout_seconds)
            )
        return self._session

    def _resolve_api_key(self) -> str:
        # An explicit key from the configuration wins over the environment.
        if self.api_key:
            return self.api_key
        api_key = os.environ.get(self.api_key_env, "").strip()
        if not api_key:
            raise LLMError(
                "no API key configured: set llm.api_key or the "
                f"environment variable {self.api_key_env}"
            )
        return api_key

    async def complete(self, system_prompt: str, user_text: str) -> str:
        api_key = self._resolve_api_key()
        payload = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_text},
            ],
            "temperature": self.temperature,
            "max_tokens": self.max_tokens,
            "stream": False,
        }
        try:
            async with self._get_session().post(
                f"{self.base_url}/chat/completions",
                json=payload,
                headers={"Authorization": f"Bearer {api_key}"},
            ) as response:
                body = await response.json(content_type=None)
                if response.status != 200:
                    message = (
                        body.get("error", {}).get("message", "")
                        if isinstance(body, dict)
                        else ""
                    )
                    raise LLMError(
                        f"LLM request failed with {response.status}: {message}"
                    )
        except aiohttp.ClientError as exc:
            raise LLMError(f"unable to reach the LLM service: {exc}") from exc
        except TimeoutError as exc:
            raise LLMError("LLM request timed out") from exc
        try:
            content = body["choices"][0]["message"]["content"]
        except (KeyError, IndexError, TypeError) as exc:
            raise LLMError("LLM returned a malformed response") from exc
        text = str(content or "").strip()
        if not text:
            raise LLMError("LLM returned an empty response")
        return text

    async def aclose(self) -> None:
        if self._session is not None and not self._session.closed:
            await self._session.close()
